### ISSUE TITLE
Feature: RedundantCondition for unnecessary !empty checks

### ISSUE THREAD

**Issue by kkmuffme** [vimeo/psalm#8507]
https://psalm.dev/r/f2049080ac

the `!empty( $foo )` should give a `RedundantConditionGivenDocblockType` error
(there are some additional cases with !empty/empty similar)

Same for:

```
if ( empty( $foo ) && empty( $foo['bar'] ) ) { // unnecessary empty( $foo ) &&
if ( !empty( $foo ) && !empty( $foo['bar'] ) ) { // unnecessary !empty( $foo ) && 
if ( empty( $foo ) || empty( $foo['bar'] ) ) { // unnecessary empty( $foo ) || 
if ( !empty( $foo ) || !empty( $foo['bar'] ) ) { // unnecessary || !empty( $foo['bar'] )

if ( empty( $foo ) && !isset( $foo['bar'] ) ) { // unnecessary && !isset( $foo['bar'] )
if ( empty( $foo ) || !isset( $foo['bar'] ) ) { // unnecessary && !isset( $foo['bar'] )
if ( empty( $foo ) && isset( $foo['bar'] ) ) {  // unnecessary && isset( $foo['bar'] )
```

**Comment by AndrolGenhald** [comment]
I don't think we support these sorts of redundancy checks at all, if a check is non-redundant at the point when it happens, but it's still redundant overall we don't seem to catch it: https://psalm.dev/r/c576174634 https://psalm.dev/r/21d498354f

I'm also not sure if we _should_ support that. At the very least, we would have to only perform that analysis when only pure functions are involved.

**Comment by AndrolGenhald** [comment]
That's different, your `isset` example is [already contradictory](https://psalm.dev/r/7e0c7aba25) even without the `$foo === "hello"` check (or [redundant](https://psalm.dev/r/7b68e4e3a3) if you remove the negation). If you change it so that it's [not redundant](https://psalm.dev/r/0bf2a870d9) you end up with the [same problem](https://psalm.dev/r/a069aad2ee), the expression isn't considered redundant because it's only redundant when considering the entire statement, rather than considering the expressions from left to right.
